### DISCUSSION TITLE
Testing custom error handlers

### DISCUSSION THREAD

**Discussion by gonzacastro** [fastapi/fastapi#6248] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class BaseError(Exception):
    def __init__(self, message, status_code) -> None:
        self.message = message
        self.status_code = status_code

async def error_handler(request: Request, exc: BaseError):
    return JSONResponse(content={"message": exc.message}, status_code=exc.status_code)
```

### Description

Given the example code, which is only a custom exception class and its handler, what I was trying while making the tests was to cover the handler lines but it keeps being uncovered, even when I force the exception to be raised. By the time, it's not that relevant because it's only one line but it'd get a pain if one handler is larger...
I tried several things, from raising the exception myself to adding the handler to the test client, so please, if anyone know about this I'd really appreciate the help.

### Operating System

Windows

### FastAPI Version

0.78.0

### Python Version

3.9.13

**Comment by odiseo0** [comment]
What tool are you using for testing the application? Also, can you show how are you testing it?

**Comment by odiseo0** [comment]
Btw, this code is similar to yours

```py
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse


class BaseError(Exception):
    def __init__(self, message: str, status_code: int) -> None:
        self.message = message
        self.status_code = status_code


app = FastAPI()

@app.exception_handler(BaseError)
async def error_handler(request: Request, exc: BaseError):
    print("Hello")
    return JSONResponse(content={"message": exc.message}, status_code=exc.status_code)


@app.get("/")
def index():
    raise BaseError(message="Error", status_code=404)
```

The output I get when visiting `/` is the following:
```console
Hello
INFO:     127.0.0.1:63867 - "GET / HTTP/1.1" 404 Not Found
```

Maybe you're missing the `exception_handler` decorator

**Comment by gonzacastro** [comment]
Thanks for replying! I think I'm not missing the decorator, actually the example code is just a snipper, the actual code is divided into different files, and the decorator is for the app. When I run the app it works and handles the exception properly but my problem is when I want to write the unit tests. Please tell me if I was clear.

**Comment by odiseo0** [comment]
Understood. How are you writing the test and what is the behavior that you expect from the `exception_handler`?

**Comment by gonzacastro** [comment]
I'm running the tests with pytest and I'm patching some random app function with a BaseError side effect so it triggers the handler. What I expect is for the test to execute the handler code lines. FYI I also tried to call the handler method directly but it returns me a coroutine object instead of the JSONResponse.

**Comment by JarroVGIT** [comment]
> I'm patching some random app function with a BaseError side effect so it triggers the handler.

I don’t think this works the way you think it works. Especially if you say that when testing the handler directly, and you didn’t expect a coroutine to be returned by an asynchronous function🤔 your (async) test client needs to await the function for it to return a value.

Provide the testcase that you tried, including any fixtures and mocking, maybe that clarifies.

**Comment by gonzacastro** [comment]
After several tries, I could achieve to cover the file with tests.
```
@pytest.mark.anyio
async def test_error_handler():
    response = await error_handler(None, error)
    assert response.body == {'message': 'test'}
```
Thank you guys for replying!

**Comment by Kludex** [comment, marked as answer]
The best way to test custom exception handlers would be to test an endpoint that triggers that exception, and verify that the response matches what the custom exception handler is doing.

**Comment by blenzi** [comment]
Hello, would you have a full working example, please? I would like to catch any exception, that will normally turn into HTTPException with status_code 500. But during testing, the original exception is raised and not handled.

```
@app.exception_handler(500)
async def server_error(request: Request, exc: Exception):
    ...
    return JSONResponse(
        status_code=500,
        content={"message": f"Problem with request, issue opened: {issue_iid}"},
    )
```

Test file:
```
import pytest

@pytest.fixture
def mock_error_500():
    """Add route to raise ValueError"""

    @app.get("/raise-500")
    async def raise_value_error():
        raise ValueError("This is a test")

def test_server_error(fixture, mock_error_500, get_headers):  # noqa F811
    with get_client() as client:
        response = client.get("/raise-500")  # raises ValueError !
        assert response.status_code == 500
```

**Reply by Vandivier** [reply]
where is get_client defined?

**Comment by tianjing-li** [comment]
@Kludex have there been any updates on this issue? I also was not able to directly test the behaviour of the exception handler, it works when running the app and an endpoint raises an unexpected error, but testing it just raises the initial error without going through the handler. 

My code is similar to @blenzi 's

**Comment by Vandivier** [comment]
this case fails even though my custom handler catches the error and replies (proven by breakpoint and integration tests):
```
@pytest.mark.asyncio
async def test_post_feedback_failure(
    test_client, feedback_payload
):
    """Test the failure case for POST /feedback endpoint without a real DB."""
    # First, validate the payload matches FeedbackRequest schema
    FeedbackRequest(**feedback_payload)

    with patch(
        "src.routes.feedback_routes.save_feedback_from_request",
        new_callable=AsyncMock,
    ) as mock_save_feedback_from_request:
        mock_save_feedback_from_request.side_effect = Exception("Database error")
        response = test_client.post("/feedback", json=feedback_payload)
        assert response is not None
```

Pytest reports:
`FAILED ...::test_post_feedback_failure - Exception: Database error`

**Reply by Vandivier** [reply]
here's a more complete minimal repro that fails in python 3.12:
```python
#######################
# temp_test.py
#######################

import pytest
from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse
from starlette.testclient import TestClient

#
# Define the FastAPI application in this same file
#

app = FastAPI()

@app.exception_handler(Exception)
async def catch_all_exceptions(request: Request, exc: Exception):
    """
    Catch-all exception handler.
    Returns HTTP 500 and a JSON body with the error detail.
    """
    return JSONResponse(
        status_code=500,
        content={"detail": str(exc)},
    )

@app.get("/error")
async def error_endpoint():
    """
    This endpoint raises a raw Python Exception("Database error").
    The catch_all_exceptions should handle it and return HTTP 500.
    """
    raise Exception("Database error")

#
# Tests in the same file
#

@pytest.fixture
def client():
    """
    Creates a fresh TestClient instance for each test.
    """
    return TestClient(app)


def test_error_endpoint(client):
    """
    Test that verifies the /error endpoint, which raises a Python exception,
    is converted by the global exception handler to an HTTP 500 response.
    """
    response = client.get("/error")

    # We expect a 500 response, not an uncaught Python exception.
    assert response.status_code == 500

    # The response JSON should contain the exception string.
    data = response.json()
    assert "detail" in data
    assert data["detail"] == "Database error"
```

**Reply by Vandivier** [reply]
I found a solution!

modify the test client this way:
```

@pytest.fixture
def client():
    """
    Creates a fresh TestClient instance for each test.
    """
    client = TestClient(app, raise_server_exceptions=False)
    return client
```

and the repro will pass

**Reply by davidhuser** [reply]
worked -- if you use the AsyncClient add `raise_app_exceptions=False` like this:

```python

    async with AsyncClient(transport=ASGITransport(app=app, raise_app_exceptions=False), base_url='http://test') as client:
        response = await client.get('/test/exception')
        assert response.status_code == 500
```

**Reply by dnk8n** [reply]
Thank you @Vandivier. Crazy that the best FastAPI docs seem to be tidbits I find buried in discussion threads.
